zlib/http.go

```diff
@@ -176,8 +177,15 @@ func (response HTTPResponse) canRedirectWithConn(conn *Conn) bool {
 		return false
 	}
 
-	relativePath := redirectUrl.Host == ""
-	matchesHost := (strings.Contains(redirectUrl.Host, targetUrl.Host)) || (redirectUrl.Host == remoteIpAddress)
+	targetHost := targetUrl.Host
+	if targetHost == "" {
+		targetHost = conn.domain
+	}
+	redirectHost := redirectUrl.Host
+
+	relativePath := redirectHost == ""
+	matchesHost := (strings.Contains(redirectHost, targetHost) && util.TLDMatches(redirectHost, targetHost)) ||
```

Comment:
@dadrian the `strings.Contains(redirectHost, targetHost)` would catch domain a.b.c being redirected to a.d.c